lib/portlayer/exec/base.go

```diff
@@ -302,6 +309,13 @@ func (c *containerBase) shutdown(ctx context.Context, waitTime *int32) error {
 			// Just warn and proceed to waiting for power state per issue https://github.com/vmware/vic/issues/5803
 			// Description above in function kill()
 			log.Warnf("%s: %s", msg, err)
+
+			// If the error tells us "The attempted operation cannot be performed in the current state (Powered off)" (InvalidPowerState),
+			// we can avoid hard poweroff (issues #6236 and #6252). Here we wait for the power state changes instead of return
+			// immediately to avoid excess vSphere queries
+			if isInvalidPowerStateError(err) {
+				killed = true
```

Comment:
Shouldn't we just return here? Why should we move on to call `WaitForPowerState`?

Reply:
When I tried `docker stop` on 10 containers in parallel, I found that the container with the `InvalidPowerState` error retried `containerProxy.stop()` several times within 8 sec. Actually more than half of the containers (out of 10) suffered from the same invalidPowerstate error. So we had too many vsphere queries. And this would be normal in production if the customer starts/stops many containers in parallel. Therefore, instead of returning immediately, I want to waitForPowerState() here to avoid the excess vsphere queries.  The performance will not be worse than the case without my code change.

Reply:
Doesn't waitForPowerState() also issue a vsphere query?

Reply:
@sflxn Yes `waitforPowerState()` is a vsphere query. In my local test, I find that `returning nil` without `WaitForPowerState()` results in way too many vsphere requests (due to the retry operations at the persona side) within several seconds, whereas we could've avoid these requests by calling `WaitForPowerState()`.